src/compiler/parser.ts

```diff
@@ -1971,8 +1971,14 @@ namespace Parser {
 
         // If we parsed this as an external module, it may contain top-level await
         if (!isDeclarationFile && isExternalModule(sourceFile) && sourceFile.transformFlags & TransformFlags.ContainsPossibleTopLevelAwait) {
+            const oldSourceFile = sourceFile;
             sourceFile = reparseTopLevelAwait(sourceFile);
-            setFields(sourceFile);
+            if (sourceFile !== oldSourceFile) {
+                setFields(sourceFile);
+                // While we are parsing the file, we dont want to use the prev source file as original node anywhere
+                // as it wont have information that gets set on the source file later by program (like paths, computed imports etc)
+                sourceFile.original = undefined;

```

Comment:
At a glance, needing to do this feels weird... My understanding is that this is needed because this is the only place in the parser that uses a `factory.update*` method.

What's interesting is that the factory has a mechanism (`NoOriginalNode`) which is set for the factory the parser uses, but it's not used consistently used inside the factory (search `setOriginalNode` in `nodeFactory.ts`). Part of me thinks that the better fix is to make sure we're always respecting that flag in the factory.

Reply:
> it's not used consistently used inside the factory (search `setOriginalNode` in `nodeFactory.ts`). Part of me thinks that the better fix is to make sure we're always respecting that flag in the factory.

I agree with @jakebailey's assessment.

Reply:
For fun, I tried replacing all uses of `setOriginalNode` with `update`, and everything passed except a bunch of source map tests; `update` also `setRange`s so not sure if the diff was good or bad (I just thought I'd try and see what happened).

Reply:
Ok there is `cloneNode` method that uses `cloneSourceFile` method and is also used by update worker. Are clone methods suppose to set original node or use `update` of the factory. I am not sure i know the protocol there to know if just removing `setOriginalNode` from `cloneSourceFile` is correct thing to do or not.

Reply:
I don't think removing it is the right thing; moreso that any use of `setOriginalNode` should be gated on the `NoOriginalNode` flag. `update` does that automatically because it is conditionally set based on the flag (then that determines if `setOriginalNode` is called), but given `update` also updates position info, it may be sufficient to just conditionalize any calls to `setOriginalNode` on the flag, or pull it out to a helper kinda like `update`.